### PR TITLE
Adds support for `ISystemReset` in test fixture

### PR DESCRIPTION
**Pull request by arjo129**
🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

# 🎉 New feature

Closes #<NUMBER>

## Summary
This PR adds support for the Reset API to the test fixture. As `TestFixture` is one of the main ways one can get access to the ECM in python when trying to write some scripts for Deep Reinforcement Learning I realized that without `Reset` supported in the `TestFixture` API, end users would have a very hard time using our python APIs (which are actually quite nice). For reference I'm hacking a demo template here (its still barebones):

https://github.com/arjo129/gz_deep_rl_experiments/tree/ionic

## Test it
I've added a test in #2648, but that includes some additional APIs to improve usability.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/TestFixture.cc [hunk 1]
```diff
@@ -200,6 +217,16 @@ TestFixture &TestFixture::OnPostUpdate(std::function<void(
...
+//////////////////////////////////////////////////
+TestFixture &TestFixture::OnReset(std::function<void(
+          const UpdateInfo &, EntityComponentManager &)> _cb)
+{
+  if (nullptr != this->dataPtr->helperSystem)
+    this->dataPtr->helperSystem->resetCallback = std::move(_cb);
+  return *this;
+}
+
+
```
Comment (ahcorde) [inline src/TestFixture.cc:229]:
```suggestion
```

**Review by ahcorde — APPROVED** [review]

**Comment by arjo129** [comment]
Upon further examination I think this may break ABI cause we are adding a base class. I probably should port this to Gazebo J.

File: python/src/gz/sim/TestFixture.cc [hunk 2]
```diff
@@ -83,6 +83,17 @@ defineSimTestFixture(pybind11::object module)
...
+  .def(
+    "on_reset", WrapCallbacks(
+      [](TestFixture* self, std::function<void(
+          const UpdateInfo &, EntityComponentManager &)> _cb)
+      {
+        self->OnReset(_cb);
+      }
+    ),
+    pybind11::return_value_policy::reference,
+    "Wrapper around a system's post-update callback"
```
Comment (iche033) [inline python/src/gz/sim/TestFixture.cc:96]:
```suggestion
    "Wrapper around a system's reset callback"
```

File: include/gz/sim/TestFixture.hh [hunk 3]
```diff
@@ -96,6 +96,12 @@ class GZ_SIM_VISIBLE TestFixture
   public: TestFixture &OnPostUpdate(std::function<void(
       const UpdateInfo &, const EntityComponentManager &)> _cb);
 
+  /// \brief Wrapper around a system's update callback
+  /// \param[in] _cb Function to be called every update
```
Comment (iche033) [inline include/gz/sim/TestFixture.hh:100]:
```suggestion
  /// \param[in] _cb Function to be called every reset
```

Comment (arjo129) [reply include/gz/sim/TestFixture.hh]: ecbe7244bff687484d6f3532179e52121fe5f1a1

**Pull request merged**